src/SourceGenerators/Uno.UI.SourceGenerators/ImplementedRoutedEvents/ImplementedRoutedEventsGenerator.cs

```diff
@@ -0,0 +1,208 @@
+﻿#nullable enable
+
+using System;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Uno.UI.SourceGenerators.Helpers;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+
+#if NETFRAMEWORK
+using Uno.SourceGeneration;
+#endif
+
+
+namespace Uno.UI.SourceGenerators.ImplementedRoutedEvents
+{
+	[Generator]
+	public class ImplementedRoutedEventsGenerator : ISourceGenerator
+	{
+		private const string ImplementedRoutedEventsPropertyName = "ImplementedRoutedEvents";
+
+		public void Initialize(GeneratorInitializationContext context)
+		{
+			// No initialization required.
+		}
+
+		public void Execute(GeneratorExecutionContext context)
+		{
+			if (!DesignTimeHelper.IsDesignTime(context))
+			{
+				var controlSymbol = context.Compilation.GetTypeByMetadataName("Windows.UI.Xaml.Controls.Control");
+				if (controlSymbol is null)
+				{
+					return;
+				}
+
+				var implementedRoutedEventsSymbol = controlSymbol.GetMembers(ImplementedRoutedEventsPropertyName).SingleOrDefault(m => m.Kind == SymbolKind.Property) as IPropertySymbol;
+				if (implementedRoutedEventsSymbol is null)
+				{
+					return;
+				}
+
+				if (!implementedRoutedEventsSymbol.ContainingAssembly.IsSameAssemblyOrHasFriendAccessTo(context.Compilation.Assembly))
+				{
+					return;
+				}
```

Comment:
This prevents assemblies that can't access Uno.UI internals from generating code because the `ImplementedRoutedEvents` property is internal.

I'm thinking of one of two enhancements here:

- Make the property public (Uno-specific API) to let user-custom controls benefit from the performance improvement; **or**
- If we can't do the first suggestion, give internals visible to Uno.UI.Maps so that `MapPresenter` can benefit from this (probably there are other assemblies that can benefit too?)

Reply:
In my opinion, it can be made public. There's no reason to restrict performance improvements to Uno only.

Complex libraries like Telerik, Syncfusion, WinToolKit should benefit from this as well.